applications/tari_console_wallet/src/ui/app.rs

```diff
@@ -0,0 +1,263 @@
+        // taking a read lock and making this method async, which adds some read/write cycles,
+        // so it's easier to just ask for it right now
+        let app_state = Arc::new(RwLock::new(AppState::new(&node_identity, network)));
+
+        let tabs = TabsContainer::<B>::new(title.clone())
+            .add("Transactions".into(), Box::new(TransactionsTab::new()))
+            .add("Send/Receive".into(), Box::new(SendReceiveTab::new()))
+            .add("Network".into(), Box::new(NetworkTab::new()));
+
+        Self {
+            title,
+            wallet,
+            should_quit: false,
+            app_state,
+            // tabs: TabsState::new(vec!["Transactions".into(), "Send/Receive".into(), "Network".into()]),
+            tabs,
+        }
+    }
+
+    pub async fn on_control_key(&mut self, c: char) {
+        match c {
+            'c' => {
+                self.should_quit = true;
+            },
+            _ => {},
+        }
+    }
+
+    pub async fn on_key(&mut self, c: char) {
+        match c {
+            'q' => {
+                self.should_quit = true;
+            },
+            '\t' => {
+                self.tabs.next();
+                return;
+            },
+            _ => {
+                let mut app_state = self.app_state.write().await;
+                self.tabs.on_key(&mut app_state, c)
+            },
+        }
+    }
+
+    pub async fn on_up(&mut self) {
+        let mut app_state = self.app_state.write().await;
+        self.tabs.on_up(&mut app_state);
+    }
+
+    pub async fn on_down(&mut self) {
+        let mut app_state = self.app_state.write().await;
+        self.tabs.on_down(&mut app_state);
+    }
+
+    pub async fn on_right(&mut self) {
+        // This currently doesn't need app_state, but is async
+        // to match others
+        self.tabs.next();
+    }
+
+    pub async fn on_left(&mut self) {
+        // This currently doesn't need app_state, but is async
+        // to match others
+        self.tabs.previous();
+    }
+
+    pub async fn on_esc(&mut self) {
+        let mut app_state = self.app_state.write().await;
+        self.tabs.on_esc(&mut app_state);
+    }
+
+    pub async fn on_backspace(&mut self) {
+        let mut app_state = self.app_state.write().await;
+        self.tabs.on_backspace(&mut app_state);
+    }
+
+    pub fn on_tick(&mut self) {}
+
+    pub async fn refresh_state(&mut self) -> Result<(), UiError> {
+        let mut pending_transactions: Vec<CompletedTransaction> = Vec::new();
+        pending_transactions.extend(
+            self.wallet
+                .write()
+                .await
+                .transaction_service
+                .get_pending_inbound_transactions()
+                .await?
+                .values()
+                .map(|t| CompletedTransaction::from(t.clone()))
+                .collect::<Vec<CompletedTransaction>>(),
```

Comment:
Same here - rather than using a RwLock - it's cheaper and more concurrent to clone the handle and use it mutably